### ISSUE TITLE
Autoreject cleans only some epochs, not all

### ISSUE THREAD

**Issue by henrikroehr** [autoreject/autoreject#256]
I tried to apply autoreject to a set of MEG data after epoching. One of my channels is very noisy (MEG 017), so I paid special attention to how autoreject handles this channel. While the majority of epochs looks quite well, a number of epochs are not repaired and are as noisy as before. I'll attach the plots of the first 10 and first 100 epochs before and after autoreject, on which this issue is quite visible.

I applied autoreject with the following code:
`ar=AutoReject()
epochs_clean = ar.fit_transform(epochs)`

Any ideas how to fix this?

First 10 epochs before autoreject:
![epochs_raw](https://user-images.githubusercontent.com/66763443/145012581-50798273-ae65-4fef-b030-62e676240899.png)
and after:
![epochs_cleaned](https://user-images.githubusercontent.com/66763443/145012579-56dea00c-bae8-4717-ab8c-6de847417663.png)
First 100 epochs before autoreject:
![epochs100_raw](https://user-images.githubusercontent.com/66763443/145012577-c07f7cf9-7d3b-4d85-94ba-a7a2f8476e53.png)
and after:
 ![epochs100_cleaned](https://user-images.githubusercontent.com/66763443/145012569-b865aea7-c674-4c87-b1bf-a8c04b171692.png)

**Comment by jasmainak** [comment]
can you share a part of your data somehow so I can investigate?

**Comment by henrikroehr** [comment]
Yes sure, I saved the raw and clean epochs to .txt files with pickle (only the first 10, as there file is quite large, but let me know if you'd like the whole file). You should be able to import the data with 

`with open('epochs.txt', 'rb') as file: 
    epochs = pickle.load(file)`

Thanks a lot for your support!
[epochs.txt](https://github.com/autoreject/autoreject/files/7676133/epochs.txt)
[epochs_clean.txt](https://github.com/autoreject/autoreject/files/7676148/epochs_clean.txt)

**Comment by jasmainak** [comment]
I will need your full data and script to reproduce. Can you share on dropbox? You can use `epochs.save`

**Comment by henrikroehr** [comment]
I just shared the dropbox folder with you. I used your @telecom-paristech.fr email address, is this the right one?

Again, thanks a lot for your help!

**Comment by jasmainak** [comment]
sorry I need to update the email address. Could you send to

mainakjas [at] gmail [dot] com

**Comment by henrikroehr** [comment]
Yes just shared it, did you receive it?
Here the link as well:
https://www.dropbox.com/sh/ivpd0pj3z49mdq5/AAAA_inRGIv12KQviekowEPva?dl=0

**Comment by henrikroehr** [comment]
Hi @jasmainak, die you receive the link and had time to take a look at it yet?

**Comment by jasmainak** [comment]
@henrikroehr I did manage to get around to this today. Here is the result of my investigations:

![image](https://user-images.githubusercontent.com/15852194/147872372-6b1c05d8-4661-4dc2-8181-89ecd880474b.png)

code: https://gist.github.com/jasmainak/05dc7c6d26592297d39a4fa25fb21d88

The crux of the problem is that cross-validation in autoreject causes all sorts of problems that could have been easily avoided by using all the data.

1) You get NaN values for certain folds because the threshold being investigated is too high for the data in that fold
2) Each fold has a different optimal threshold value which is different from the optimal threshold value for the entire data which results in missed epochs.

One of the reviewers of my PhD defense also pointed out that the cross-validation was completely unnecessary and I tend to agree but for any changes to be introduced to autoreject, we must validate the change with respect to a large dataset and it must pass the muster of @agramfort 

On average, autoreject probably doesn't do any worse than the strategy using all the data but it causes such glaring misses. I can only suggest one thing for now. You can supply your own `thresh_func` to `Autoreject` that builds upon the code I shared. Might need some work. If you're up for it, give it a shot and I'll try to help when you're stuck. The code to get started is here:

https://github.com/autoreject/autoreject/blob/e24eea35948d0db0ccf233d519128d8ff8aa06d2/autoreject/autoreject.py#L432-L453

**Comment by agramfort** [comment]
hum....

not sure what you mean with no cross-val but if you mean fitting and
testing on the same data you can use
cv object with a scikit-learn API that outputs the full list of indices for
the train and the test. We could expose this with
passing cv=1 in the public API. Wdyt? clear?

**Comment by jasmainak** [comment]
I tried `cv=1` but sklearn was complaining. Do you know if it's allowed?

**Comment by agramfort** [comment]
it would not be allowed with sklearn but you can do if cv=1 use a splitter like this:

```
import numpy as np
from sklearn.model_selection import BaseCrossValidator

class NoSplitter(BaseCrossValidator):
    def __init__(self):
        pass

    def get_n_splits(self, X=None, y=None, groups=None):
        return 1

    def split(self, X, y, groups=None):
        idx = np.arange(len(X))
        yield idx, idx

if __name__ == '__main__':
    cv = NoSplitter()
    X = np.random.randn(3, 2)
    for train, test in cv.split(X, None):
        print(train, test)
```

**Comment by jasmainak** [comment]
excellent! Yes that's what I was looking for. In that case, we can add a `cv` argument to `compute_thresholds` that by default does `cv=10` but can also be `cv=NoSplitter()`. So the code would be something like

```py
from functools import partial

my_thresh_func = compute_thresholds(..., cv=NoSplitter())
Autoreject(thresh_func=compute_threshold_full)
```

Wdyt @henrikroehr ?

**Comment by henrikroehr** [comment]
Hey apologies for my late response. I think this solves my problem, thanks a lot!

**Comment by henrikroehr** [comment]
Hi @jasmainak,

Apologies for reopening this thread after long time. At first it looked like the code you and @agramfort provided to me fixed the issue, and it mostly did, i.e almost all the bad epochs were repaired. However, now I found another epoch which eluded autoreject, and I was wondering if you have any idea what might be the cause for it.

The raw data are still the same as in the dropbox I shared with you (see comment above). I adjusted the autoreject.py file by replace the value of cv in the _compute_thresholds() function with NoSplitter() as agramford suggested. My own code and the adjustments I made to the autoreject.py file are below. The bad epoch is epoch 148 in channel 17, which can be found with epochs_clean[126].

Thanks again for your support!

My code:
`import os
import numpy as np
import mne
from autoreject import (AutoReject, compute_thresholds)
from mne.preprocessing import (ICA, create_eog_epochs, create_ecg_epochs, corrmap)

## Load data
os.chdir('/home/henrik/research/Data/SB10')

#import cleaned data
raw = mne.io.read_raw_bti(os.path.join('cleaned', os.listdir('cleaned')[0]),
                          config_fname='raw/config', 
                          head_shape_fname='raw/hs_file')

raw.crop(tmax=600).load_data()


## find events of transitioning between experimental modes
events = mne.find_events(raw, stim_channel='STI 014', output='step', min_duration=1) #min_duration is set to 1 second to filter out fluctuations

event_dict = {'rest': 100, 'boundary/dissolution': 150, 'boundary/maintenance': 200}


## As the data are not event-related, they are divided into regular time spans
## This is done for each experimental mode separately to preserve the ids
events_epochs = np.tile(0,(1,3)) # Create dummy array. Necessary to join epochs of different experimental modes in a single data structure. Will be deleted at the end.
for i in range(0, len(events), 2): # For each step/experimental mode
    beginning = events[i][0]/1000 # Divide by 1000 because mne.make_fixed_length_events takes input in seconds
    end = events[i+1][0]/1000
    events_new = mne.make_fixed_length_events(raw, id=int(events[i][2]), start=beginning, stop=end, duration=1, overlap=0.25) #this is identical to events of duration 0.75 without overlap, as only the start time is recorded. However, the duration will be specified when epoching.
    events_epochs = np.append(events_epochs,events_new,axis=0) # append events
events_epochs = np.delete(events_epochs,0,axis=0) # delete dummy array

epochs = mne.Epochs(raw, events_epochs, event_id=event_dict, tmin=0, tmax=1, baseline=None, preload=True)

del raw #free up memory

## automatically repair bad epochs
thresh_func = compute_thresholds(epochs=epochs)
ar = AutoReject(thresh_func=thresh_func)
epochs_clean = ar.fit_transform(epochs)

del epochs #free up memory

epochs_clean[125:128].plot()`

Adjustments to autoreject.py were made after line 417 (marked by ### inserted ### or ### changed ###):
`### inserted ###
class NoSplitter(BaseCrossValidator):
    def __init__(self):
        pass

    def get_n_splits(self, X=None, y=None, groups=None):
        return 1

    def split(self, X, y=None, groups=None):
        idx = np.arange(len(X))
        yield idx, idx
### inserted


def _compute_thresholds(epochs, method='bayesian_optimization',
                        random_state=None, picks=None, augment=True,
                        dots=None, verbose=True, n_jobs=1):
    if method not in ['bayesian_optimization', 'random_search']:
        raise ValueError('`method` param not recognized')
    picks = _handle_picks(info=epochs.info, picks=picks)
    _check_data(epochs, picks, verbose=verbose,
                ch_constraint='data_channels')
    picks_by_type = _get_picks_by_type(picks=picks, info=epochs.info)
    picks_by_type = None if len(picks_by_type) == 1 else picks_by_type  # XXX
    if picks_by_type is not None:
        threshes = dict()
        for ch_type, this_picks in picks_by_type:
            threshes.update(_compute_thresholds(
                epochs=epochs, method=method, random_state=random_state,
                picks=this_picks, augment=augment, dots=dots,
                verbose=verbose, n_jobs=n_jobs))
    else:
        n_epochs = len(epochs)
        data, y = epochs.get_data(), np.ones((n_epochs, ))
        if augment:
            epochs_interp = _clean_by_interp(epochs, picks=picks,
                                             dots=dots, verbose=verbose)
            # non-data channels will be duplicate
            data = np.concatenate((epochs.get_data(),
                                   epochs_interp.get_data()), axis=0)
            y = np.r_[np.zeros((n_epochs, )), np.ones((n_epochs, ))]
### changed
        # cv = StratifiedShuffleSplit(n_splits=10, test_size=0.2,
        #                             random_state=random_state)
        
        if __name__ == '__main__':
            cv = NoSplitter()
            #X = np.random.randn(3, 2)
            # for train, test in cv.split(X, None):
            #     print(train, test)
        else:
            cv = 10
### changed ###`

**Comment by jasmainak** [comment]
You'll need to update autoreject to the latest development version. But before you do that, you should perhaps push your changes in autoreject somewhere so we can copy it into the main repository. Are you using version control? Once you do that, just roll back the current edit and get the latest copy. I'll then work on incorporating your changes into the main repository and you can update your autoreject once they are incorporated. How does that sound to you?

**Comment by henrikroehr** [comment]
Okay sounds good. I though I had the latest version, but now noticed it was recently updated.

I pushed the autoreject.py file with the changes I made to https://github.com/henrikroehr/autoreject_changed

**Comment by jasmainak** [comment]
Thank you! I will take a look soon :)
